test/frontend/frontend.test.js

```diff
@@ -0,0 +1,353 @@
+          })
+        })
+        
+      }
+
+      readRaw.readAsArrayBuffer(newFile);
+    })
+  })
+
+});
+
+describe('File Receiver', function() {
+
+  class FakeXHR {
+    constructor() {
+      this.response = file;
+      this.status = 200;
+    }
+
+    static setup() {
+      FakeXHR.prototype.open = sinon.spy();
+      FakeXHR.prototype.send = function () {
+        this.onload();
+      }
+
+      FakeXHR.prototype.originalXHR = window.XMLHttpRequest;
+
+      FakeXHR.prototype.getResponseHeader = function () {
+        return JSON.stringify({
+          aad: arrayToHex(new Uint8Array(fileHash)),
+          filename: 'hello_world.txt',
+          id: encryptedIV
+        })
+      }
+      window.XMLHttpRequest = FakeXHR;
+    }
+
+    static restore() {
+      // originalXHR is a sinon FakeXMLHttpRequest, since
+      // fakeServer.create() is called in frontend.bundle.js
+      window.XMLHttpRequest.prototype.originalXHR.restore();
+    }
+  }
+  
+  const cb = function(done) {
+    if (file === undefined ||
+        encryptedIV === undefined || 
+        fileHash === undefined || 
+        secretKey === undefined) {
+      assert.fail('Please run file sending tests before trying to receive the files.');
+      done();
+    }
+
+    FakeXHR.setup();
+    done();
+  }
+
+  before(cb)
+
+  after(function() {
+    FakeXHR.restore();
+  })
+
+  it('Should decrypt properly', function() {
+    const fr = new FileReceiver();
+    location.hash = secretKey;
+    return fr.download().then(([decrypted, name]) => {
+      assert(name);
+      assert(new Uint8Array(decrypted).toString() ===
+                  new Uint8Array(originalBlob).toString())
+    }).catch(err => {
+      console.log(err, err.stack);
+      assert.fail();
+    })
+  })
+
+  it('Should emit decrypting events', function() {
+    const fr = new FileReceiver();
+    location.hash = secretKey;
+
+    let testDecrypting = true;
+
+    fr.on('decrypting', isStillDecrypting => {
+      assert(!(!testDecrypting && isStillDecrypting));
+      testDecrypting = isStillDecrypting;
+    });
+
+    fr.on('safe', isSafe => {
+      assert(isSafe);
+    })
+
+    return fr.download().then(([decrypted, name]) => {
+      assert(decrypted);
+      assert(name);
+      assert(!testDecrypting);
+    }).catch(err => {
+      console.log(err, err.stack);
+      assert.fail();
+    })
+  })
+
+  it('Should emit hashing events', function() {
+    const fr = new FileReceiver();
+    location.hash = secretKey;
+
+    let testHashing = true;
+
+    fr.on('hashing', isStillHashing => {
+      assert(!(!testHashing && isStillHashing));
+      testHashing = isStillHashing;
+    });
+
+    fr.on('safe', isSafe => {
+      assert(isSafe);
+    })
+
+    return fr.download().then(([decrypted, name]) => {
+      assert(decrypted);
+      assert(name);
+      assert(!testHashing);
+    }).catch(err => {
+      assert.fail();
+    })
+  })
+
+  it('Should catch tampered checksums', function(done) {
```

Comment:
I think "tampered" isn't the best word for this. It took me a while to understand what you meant since tampering can happen at other times besides creation. Phony, bogus, or fraudulent better imply it's the creator that's being a jerk.